astropy/units/decorators.py

```diff
@@ -5,10 +5,85 @@
 
 from ..utils.decorators import wraps
 from ..utils.compat import funcsigs
+from ..utils.misc import isiterable
 
 from .core import Unit, UnitsError, add_enabled_equivalencies
 from .physical import _unit_physical_mapping
 
+def _get_allowed_units(targets):
+    """
+    From a list of target units (either as strings or unit objects) and physical
+    types, return a list of Unit objects.
+    """
+
+    allowed_units = []
+    for target in targets:
+        if isinstance(target, str): # Could be a string unit or physical type
```

Comment:
With my comment below, where I suggest dealing with `None` in the main code, we can insist all targets are either `Unit` or `string, and this `isinstance` check can be removed (since `Unit` instances pass unchanged through the `Unit` initlializer -- and this is very fast).